tidymodels 2024 Q1 roundup post (#694)
tidyverse/tidyverse.org: content/blog/tidymodels-2024-q1/index.Rmd

-   [Survival analysis for time-to-event data with tidymodels](https://www.tidyverse.org/blog/2024/04/tidymodels-survival-analysis/)
-   [Fair machine learning with tidymodels](https://www.tidyverse.org/blog/2024/03/tidymodels-fairness/)
- TODO: link to release post for tune 1.2.0

Since [our last roundup post](https://www.tidyverse.org/blog/2024/01/tidymodels-2023-q4/), there have been CRAN releases of `r nrow(tm_data)` tidymodels packages. Here are links to their NEWS files:

```{r}
#| echo: FALSE
#| results: asis

cat(tm_data$news, sep = "\n")
```

We'll highlight a few especially notable changes below: new prediction options in censored, consistency in augmenting parsnip models and workflows, as well as a new autoplot type for workflow sets.

```{r}
#| message: FALSE
library(tidymodels)
library(censored)
```

## New prediction options in censored

> As part of the framework-wide integration of survival analysis, also censored, the parsnip extension package for censored regression, received some love in the form of new prediction options.

Reviewer: ```suggestion
As part of the framework-wide integration of survival analysis, censored, the parsnip extension package for censored regression, received some love in the form of new prediction options.
```

It'd be a loss of context, but I might even just transition to "the parsnip extension package censored" here.

> Random forest with the  `"aorsf"` engine can now predict survival time, thanks to the new feature in the [aorsf](https://docs.ropensci.org/aorsf/) package itself. This means that now all engines in censored can predict survival time.

Reviewer: ```suggestion
Random forests with the  `"aorsf"` engine can now predict survival time, thanks to the new feature in the [aorsf](https://docs.ropensci.org/aorsf/) package itself. This means that all engines in censored can now predict survival time.
```

Let's predict survival time for the first five rows of the lung cancer dataset, survival analysis' `mtcars`. 

```{r aorsf}
rf_spec <- rand_forest() |>
  set_engine("aorsf") |>
  set_mode("censored regression")

rf_fit <- rf_spec |>
  fit(Surv(time, status) ~ age + sex, data = lung)

lung_5 <- lung[1:5, ]
predict(rf_fit, new_data = lung_5, type = "time")
```

> Some models allow for predictions based on different values for tuning parameter without having to refit the model. In parsnip, we refer to this as ["the submodel trick"](https://parsnip.tidymodels.org/articles/Submodels.html). Some of those models are regularized models fitted with the [glmnet](https://glmnet.stanford.edu/) engine. In censored, the corresponding `multi_predict()` method has now gained the prediction types `"time"` and `"raw"` in addition to the existing types `"survival"` and `"linear_pred"`. 

Reviewer: ```suggestion
Some models allow for predictions based on different values for tuning parameter without having to refit the model. In parsnip, we refer to this as ["the submodel trick."](https://parsnip.tidymodels.org/articles/Submodels.html) Some of those models are regularized models fitted with the [glmnet](https://glmnet.stanford.edu/) engine. In censored, the corresponding `multi_predict()` method has now gained the prediction types `"time"` and `"raw"` in addition to the existing types `"survival"` and `"linear_pred"`. 
```

More North American Attack of the English Language, feel free to ignore if you so please.

Let's fit a regularized Cox model to illustrate. Note how we set the `penalty` to a fixed value of `0.1`.

```{r coxnet-fit}
cox_fit <- proportional_hazards(penalty = 0.1) |>
  set_engine("glmnet") |>
  set_mode("censored regression") |>
  fit(Surv(time, status) ~ ., data = lung)
```

Predictions made with `predict()` use that penalty value of 0.1. With `multi_predict()`, we can change that value to something different without having to refit. Conveniently, we can predict for multiple penalty values as well.

```{r coxnet-pred}
predict(cox_fit, new_data = lung_5, type = "time")

mpred <- multi_predict(cox_fit, new_data = lung_5, type = "time", 
                       penalty = c(0.01, 0.1)) 
mpred
```

The resulting tibble is nested by observation to follow the convention of one row per observation. For each observation, the predictions are stored in a tibble containing the penalty value along with the prediction.

```{r coxnet-pred-detail}
mpred$.pred[[2]]
```

> You can see that the predicted value from `predict()` matches the predicted value from `multi_predict()` with a penalty of 0.1.

Reviewer 2: Shouldn't this be `0.01`? Since we are showing the submodel trick

Reviewer 2: and if it isn't, I think we should show the `0.01` prediction 😄

Author: We are showing both! (Are you looking only at the source or also the rendered version?)

I mean that you can see the penalty values for the multi-predict output, find the one that says penalty = 0.1 and see that the (prediction) value is the same as the one from predict() -- because that is the penalty value we specified for the fit.  

If that is very confusing right now, I'm happy about suggestions!

Reviewer 2: sorry, I was only looking at source! then this is good!

## Consistent `augment()` for workflows and parsnip models

> If you are interested in exploring predictions in relation to predictors, `augment()` is your extened `predict()` method: it will augment the dataset with its predictions. For classification, it will add hard class predictions as well as class probabilities. For regression, it will add the numeric prediction. If the outcome variable is part of the dataset, it also calculates residuals. This has already been the case for fitted parsnip models, and now also the `augment()` method for workflows calculates residuals. 

Reviewer: ```suggestion
If you are interested in exploring predictions in relation to predictors, `augment()` is your extended `predict()` method: it will augment the inputted dataset with its predictions. For classification, it will add hard class predictions as well as class probabilities. For regression, it will add the numeric prediction. If the outcome variable is part of the dataset, it also calculates residuals. This has already been the case for fitted parsnip models, and the `augment()` method for workflows will now also calculate residuals. 
```

Reviewer: Only class probabilities when an engine supports, I think? Though that's maybe more technical than this post needs.

Author: Right, class probabilities depend on the engine! I'd leave it as is, though, unless anybody else thinks that this should be mentioned.

```{r augment}
spec_fit <- fit(linear_reg(), mpg ~ ., mtcars)
wflow_fit <- workflow(mpg ~ ., linear_reg()) %>% fit(mtcars)

augment(spec_fit, mtcars)

augment(wflow_fit, mtcars)
```

> Both methods also appened on the left-hand side of the data frame, rather than the right-hand side. This means that prediction columns are always visible when printed, even for data frames with many columns.

Reviewer: ```suggestion
Both methods also append on the left-hand side of the data frame, rather than the right-hand side. This means that prediction columns are always visible when printed, even for data frames with many columns.
```

> Both methods also append on the left-hand side of the data frame, rather than the right-hand side. This means that prediction columns are always visible when printed, even for data frames with many columns.

Reviewer 2: One thing we also guarantee is that the prediction types are in the same order between {workflows} and {parsnip}

## New autoplot type for workflow sets

Many tidymodels objects have `autoplot()` methods for quickly getting a sense of the most important aspects of an object. For workflow sets, the method shows the value of the calcualted performance metrics, as well as the respective rank of each workflow in the set. Let's put together a workflow set on the actual `mtcars` data and take a look at the default autoplot.

```{r workflowsets-autoplot}
mt_rec <- recipe(mpg ~ ., mtcars)
mt_rec2 <- mt_rec |> step_normalize(all_numeric_predictors())
mt_rec3 <- mt_rec |> step_YeoJohnson(all_numeric_predictors())

wflow_set <- workflow_set(
  list(plain = mt_rec, normalize = mt_rec2, yeo_johnson = mt_rec3), 
  list(linear_reg())
)

set.seed(1)
wflow_set_fit <- workflow_map(
  wflow_set, 
  "fit_resamples", 
  resamples = bootstraps(mtcars), 

>   seed = 1

Reviewer 2: is this overkill since we have `set.seed(1)` right before?

)

autoplot(wflow_set_fit)
```

> This allows you to grasp the metric values and rank of a workflow and let's you distinguish the type of preprocessor and model. In our case, we only have one type of model, and even just one type of preprocessor, a recipe. What we are much more interested in is which recipe corresponds to which rank. The new option of `type = "wflow_id"` let's us see which values and ranks correspond with which workflow and thus also with which recipe.

Reviewer: ```suggestion
This allows you to grasp the metric values and rank of a workflow and let's you distinguish the type of preprocessor and model. In our case, we only have one type of model, and even just one type of preprocessor, a recipe. What we are much more interested in is which recipe corresponds to which rank. The new option of `type = "wflow_id"` lets us see which values and ranks correspond with which workflow and thus also with which recipe.
```
